### PR TITLE
Support saving and loading stateful objects in Fabric

### LINKED ISSUE #18493: Saving learning rate schedules with Fabric
### Bug description

It is unclear to me how learning rate schedules should be used alongside `fabric.save`. I've defaulted to manually saving with `fabric.save(..., {..., 'schedule': schedule.state_dict()}` and manually loading with `ckpt=fabric.load(...); sched.load(ckpt['schedule'])`. As stated in #18482, the recommended way of loading is using a state object.  This assumes you pass `sched` to the `fabric.save`, not `sched.state_dict()`.  This leads to some weird behavior, see code below. 
 
It is also unclear to me whether a learning rate schedule object should be given the bare optimizer, or the wrapped Fabric optimizer object.

### What version are you seeing the problem on?

v2.0

### How to reproduce the bug

```python
import lightning
import torch

INSTANTIATE_SCHEDULE_ON_WRAPPER = False


class Network(torch.nn.Module):
    def __init__(self):
        super().__init__()

        self.fc1 = torch.nn.Linear(1000, 10)

    def forward(self, x):
        return torch.nn.functional.relu(self.fc1(x))


def no_schedule():
    fabric = lightning.Fabric(accelerator="gpu", devices=1)
    fabric.launch()

    network = Network()
    opt = torch.optim.Adam(network.parameters())

    network, opt = fabric.setup(network, opt)

    state = {"network": network, "opt": opt}
    fabric.save("no_schedule.ckpt", state)
    print("no schedule succeeded")


def exp_decay():
    fabric = lightning.Fabric(accelerator="gpu", devices=1)
    fabric.launch()

    network = Network()
    opt = torch.optim.Adam(network.parameters())

    if INSTANTIATE_SCHEDULE_ON_WRAPPER:
        network, opt = fabric.setup(network, opt)
        lr_schedule = torch.optim.lr_scheduler.ExponentialLR(opt, 0.99)
    else:
        lr_schedule = torch.optim.lr_scheduler.ExponentialLR(opt, 0.99)
        network, opt = fabric.setup(network, opt)

    state = {"network": network, "opt": opt, "lr_schedule": lr_schedule}
    fabric.save("exp_decay.ckpt", state)
    print("exp decay succeeded")


def cyclic():
    fabric = lightning.Fabric(accelerator="gpu", devices=1)
    fabric.launch()

    network = Network()
    opt = torch.optim.Adam(network.parameters())

    if INSTANTIATE_SCHEDULE_ON_WRAPPER:
        network, opt = fabric.setup(network, opt)
        lr_schedule = torch.optim.lr_scheduler.CyclicLR(
            opt, 1e-5, 1e-6, cycle_momentum=False
        )
    else:
        lr_schedule = torch.optim.lr_scheduler.CyclicLR(
            opt, 1e-5, 1e-6, cycle_momentum=False
        )
        network, opt = fabric.setup(network, opt)

    state = {"network": network, "opt": opt, "lr_schedule": lr_schedule}
    fabric.save("cyclic.ckpt", state)
    print("cyclic LR succeeded")


def main():
    print(lightning.__version__)
    torch.manual_seed(123)

    no_schedule()
    exp_decay()
    cyclic()


if __name__ == "__main__":
    main()
```

### Error messages and logs

If run with `INSTANTIATE_SCHEDULE_ON_WRAPPER = False`, the methods `no_schedule()` and `exp_decay()` succeed, but `cyclic` fails with:

```
Traceback (most recent call last):
  File "/home/nvaessen/phd/repo/nanow2v2/playground/loading_bug/buggy_saving.py", line 85, in <module>
    main()
  File "/home/nvaessen/phd/repo/nanow2v2/playground/loading_bug/buggy_saving.py", line 81, in main
    cyclic()
  File "/home/nvaessen/phd/repo/nanow2v2/playground/loading_bug/buggy_saving.py", line 71, in cyclic
    fabric.save("cyclic.ckpt", state)
  File "/home/nvaessen/phd/repo/nanow2v2/.venv/lib/python3.10/site-packages/lightning/fabric/fabric.py", line 633, in save
    return self._strategy.save_checkpoint(path=path, state=_unwrap_objects(state))
  File "/home/nvaessen/phd/repo/nanow2v2/.venv/lib/python3.10/site-packages/lightning/fabric/strategies/strategy.py", line 241, in save_checkpoint
    self.checkpoint_io.save_checkpoint(checkpoint=state, path=path, storage_options=storage_options)
  File "/home/nvaessen/phd/repo/nanow2v2/.venv/lib/python3.10/site-packages/lightning/fabric/plugins/io/torch_io.py", line 59, in save_checkpoint
    _atomic_save(checkpoint, path)
  File "/home/nvaessen/phd/repo/nanow2v2/.venv/lib/python3.10/site-packages/lightning/fabric/utilities/cloud_io.py", line 72, in _atomic_save
    torch.save(checkpoint, bytesbuffer)
  File "/home/nvaessen/phd/repo/nanow2v2/.venv/lib/python3.10/site-packages/torch/serialization.py", line 441, in save
    _save(obj, opened_zipfile, pickle_module, pickle_protocol)
  File "/home/nvaessen/phd/repo/nanow2v2/.venv/lib/python3.10/site-packages/torch/serialization.py", line 653, in _save
    pickler.dump(obj)
TypeError: cannot pickle 'WeakMethod' object
```

If run with `INSTANTIATE_SCHEDULE_ON_WRAPPER = True`, both `exp_decay` and `cyclic` fail with 

```
```Traceback (most recent call last):
  File "/home/nvaessen/phd/repo/nanow2v2/playground/loading_bug/buggy_saving.py", line 83, in <module>
    main()
  File "/home/nvaessen/phd/repo/nanow2v2/playground/loading_bug/buggy_saving.py", line 79, in main
    cyclic()
  File "/home/nvaessen/phd/repo/nanow2v2/playground/loading_bug/buggy_saving.py", line 69, in cyclic
    fabric.save("cyclic.ckpt", state)
  File "/home/nvaessen/phd/repo/nanow2v2/.venv/lib/python3.10/site-packages/lightning/fabric/fabric.py", line 633, in save
    return self._strategy.save_checkpoint(path=path, state=_unwrap_objects(state))
  File "/home/nvaessen/phd/repo/nanow2v2/.venv/lib/python3.10/site-packages/lightning/fabric/strategies/strategy.py", line 241, in save_checkpoint
    self.checkpoint_io.save_checkpoint(checkpoint=state, path=path, storage_options=storage_options)
  File "/home/nvaessen/phd/repo/nanow2v2/.venv/lib/python3.10/site-packages/lightning/fabric/plugins/io/torch_io.py", line 59, in save_checkpoint
    _atomic_save(checkpoint, path)
  File "/home/nvaessen/phd/repo/nanow2v2/.venv/lib/python3.10/site-packages/lightning/fabric/utilities/cloud_io.py", line 72, in _atomic_save
    torch.save(checkpoint, bytesbuffer)
  File "/home/nvaessen/phd/repo/nanow2v2/.venv/lib/python3.10/site-packages/torch/serialization.py", line 441, in save
    _save(obj, opened_zipfile, pickle_module, pickle_protocol)
  File "/home/nvaessen/phd/repo/nanow2v2/.venv/lib/python3.10/site-packages/torch/serialization.py", line 653, in _save
    pickler.dump(obj)
_pickle.PicklingError: Can't pickle <class 'lightning.fabric.wrappers.FabricAdam'>: attribute lookup FabricAdam on lightning.fabric.wrappers failed 
```

If we change `state={..., "lr_schedule": lr_schedule}` to `state={..., "lr_schedule": lr_schedule.state_dict()}`, saving succeeds for both options of `INSTANTIATE_SCHEDULE_ON_WRAPPER`. Using `state_dict()` directly means that calls to `schedule.step()` will not be registered in a central state object. 

### Environment

<details>
  <summary>Current environment</summary>

* CUDA:
	- GPU:
		- NVIDIA GeForce GTX 1080 Ti
	- available:         True
	- version:           11.7
* Lightning:
	- lightning:         2.0.8
	- lightning-cloud:   0.5.37
	- lightning-utilities: 0.8.0
	- pytorch-lightning: 2.0.3
	- torch:             2.0.1
	- torch-tb-profiler: 0.4.1
	- torchaudio:        2.0.1
	- torchdata:         0.6.1
	- torchmetrics:      0.11.4
	- torchvision:       0.15.1
* Packages:
	- absl-py:           1.4.0
	- aiohttp:           3.8.4
	- aiosignal:         1.3.1
	- alembic:           1.11.1
	- antlr4-python3-runtime: 4.9.3
	- anyio:             3.7.0
	- appdirs:           1.4.4
	- arrow:             1.2.3
	- async-timeout:     4.0.2
	- attrs:             23.1.0
	- autopage:          0.5.1
	- backoff:           2.2.1
	- beautifulsoup4:    4.12.2
	- blessed:           1.20.0
	- cachetools:        5.3.1
	- certifi:           2023.5.7
	- cffi:              1.15.1
	- charset-normalizer: 3.1.0
	- click:             8.1.3
	- cliff:             4.3.0
	- cloudpickle:       2.2.1
	- cmaes:             0.9.1
	- cmake:             3.26.4
	- cmd2:              2.4.3
	- colorlog:          6.7.0
	- contourpy:         1.1.0
	- croniter:          1.3.15
	- cycler:            0.11.0
	- dateutils:         0.6.12
	- deepdiff:          6.3.0
	- docker-pycreds:    0.4.0
	- exceptiongroup:    1.1.1
	- fastapi:           0.98.0
	- filelock:          3.12.2
	- fonttools:         4.40.0
	- frozenlist:        1.3.3
	- fsspec:            2023.6.0
	- gitdb:             4.0.10
	- gitpython:         3.1.31
	- google-auth:       2.20.0
	- google-auth-oauthlib: 0.4.6
	- greenlet:          2.0.2
	- grpcio:            1.54.2
	- h11:               0.14.0
	- huggingface-hub:   0.15.1
	- hydra-core:        1.3.2
	- hydra-optuna-sweeper: 1.2.0
	- hydra-submitit-launcher: 1.2.0
	- idna:              3.4
	- importlib-metadata: 6.7.0
	- iniconfig:         2.0.0
	- inquirer:          3.1.3
	- itsdangerous:      2.1.2
	- jinja2:            3.1.2
	- jiwer:             3.0.0
	- joblib:            1.2.0
	- kiwisolver:        1.4.4
	- lightning:         2.0.8
	- lightning-cloud:   0.5.37
	- lightning-utilities: 0.8.0
	- lit:               16.0.6
	- mako:              1.2.4
	- markdown:          3.4.3
	- markdown-it-py:    3.0.0
	- markupsafe:        2.1.3
	- matplotlib:        3.7.1
	- mdurl:             0.1.2
	- mpmath:            1.3.0
	- multidict:         6.0.4
	- nanow2v2:          1.0
	- networkx:          3.1
	- numpy:             1.25.0
	- nvidia-cublas-cu11: 11.10.3.66
	- nvidia-cuda-cupti-cu11: 11.7.101
	- nvidia-cuda-nvrtc-cu11: 11.7.99
	- nvidia-cuda-runtime-cu11: 11.7.99
	- nvidia-cudnn-cu11: 8.5.0.96
	- nvidia-cufft-cu11: 10.9.0.58
	- nvidia-curand-cu11: 10.2.10.91
	- nvidia-cusolver-cu11: 11.4.0.1
	- nvidia-cusparse-cu11: 11.7.4.91
	- nvidia-nccl-cu11:  2.14.3
	- nvidia-nvtx-cu11:  11.7.91
	- oauthlib:          3.2.2
	- omegaconf:         2.3.0
	- optuna:            2.10.1
	- ordered-set:       4.1.0
	- packaging:         23.1
	- pandas:            2.0.2
	- pathtools:         0.1.2
	- pbr:               5.11.1
	- pillow:            9.5.0
	- pip:               23.1.2
	- pluggy:            1.2.0
	- polars:            0.18.3
	- prettytable:       3.8.0
	- protobuf:          4.23.3
	- psutil:            5.9.5
	- pyasn1:            0.5.0
	- pyasn1-modules:    0.3.0
	- pycparser:         2.21
	- pydantic:          1.10.9
	- pygments:          2.15.1
	- pyjwt:             2.7.0
	- pyparsing:         3.1.0
	- pyperclip:         1.8.2
	- pytest:            7.4.0
	- python-dateutil:   2.8.2
	- python-dotenv:     1.0.0
	- python-editor:     1.0.4
	- python-multipart:  0.0.6
	- pytorch-lightning: 2.0.3
	- pytz:              2023.3
	- pyyaml:            6.0
	- rapidfuzz:         2.13.7
	- readchar:          4.0.5
	- regex:             2023.6.3
	- requests:          2.31.0
	- requests-oauthlib: 1.3.1
	- rich:              13.4.2
	- rsa:               4.9
	- safetensors:       0.3.1
	- scikit-learn:      1.2.2
	- scipy:             1.10.1
	- seaborn:           0.12.2
	- sentry-sdk:        1.25.1
	- setproctitle:      1.3.2
	- setuptools:        65.5.0
	- six:               1.16.0
	- smmap:             5.0.0
	- sniffio:           1.3.0
	- soundfile:         0.12.1
	- soupsieve:         2.4.1
	- sqlalchemy:        2.0.17
	- starlette:         0.27.0
	- starsessions:      1.3.0
	- stevedore:         5.1.0
	- submitit:          1.4.5
	- sympy:             1.12
	- tensorboard:       2.12.0
	- tensorboard-data-server: 0.7.1
	- tensorboard-plugin-wit: 1.8.1
	- threadpoolctl:     3.1.0
	- tokenizers:        0.13.3
	- tomli:             2.0.1
	- torch:             2.0.1
	- torch-tb-profiler: 0.4.1
	- torchaudio:        2.0.1
	- torchdata:         0.6.1
	- torchmetrics:      0.11.4
	- torchvision:       0.15.1
	- tqdm:              4.65.0
	- traitlets:         5.9.0
	- transformers:      4.30.2
	- triton:            2.0.0
	- typing-extensions: 4.6.3
	- tzdata:            2023.3
	- urllib3:           1.26.16
	- uvicorn:           0.22.0
	- wandb:             0.15.5
	- wcwidth:           0.2.6
	- websocket-client:  1.6.0
	- websockets:        11.0.3
	- werkzeug:          2.3.6
	- wheel:             0.40.0
	- yarl:              1.9.2
	- zipp:              3.15.0
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- ELF
	- processor:         
	- python:            3.10.12
	- release:           6.4.12-arch1-1
	- version:           #1 SMP PREEMPT_DYNAMIC Thu, 24 Aug 2023 00:38:14 +0000

</details>

### More info

_No response_

cc @awaelchli @carmocca @justusschock

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #18493

Sofar we've handled `state_dict()` and `load_state_dict()` automatically for the user on the module and optimizer(s) when saving and loading a state with Fabric. This PR extends this to any object in the state that is "stateful", for example learning rate schedulers:

```py
state = {"model": ..., "lr_scheduler": my_torch_scheduler}

# auomatically saves my_torch_scheduler.state_dict() now
fabric.save(path, state)

# auomatically calls my_torch_scheduler.load_state_dict() now
fabric.load(path, state)
```

cc @borda @justusschock @awaelchli @carmocca

### CODE REVIEW

File: src/lightning/fabric/strategies/strategy.py [hunk 1]
```diff
@@ -399,6 +399,8 @@ def _convert_stateful_objects_in_state(
                 converted = self.get_module_state_dict(module=obj)
             elif isinstance(obj, Optimizer):
                 converted = self.get_optimizer_state(optimizer=obj)
+            elif isinstance(obj, _Stateful):
+                converted = obj.state_dict()
```
Comment (awaelchli) [inline src/lightning/fabric/strategies/strategy.py:403]: Note: The load_checkpoint logic already handles `_Stateful`, so the only change needed is here in the saving logic.

File: src/lightning/fabric/utilities/load.py [hunk 2]
```diff
@@ -203,3 +203,21 @@ def _load_tensor(t: _NotYetLoadedTensor) -> Tensor:
...
+
+def _take_state_and_load_stateful(
+    source: Dict[str, Any], destination: Dict[str, Union[Any, _Stateful]], keys: Optional[Set[str]] = None
+) -> None:
+    """Takes the state from the source destination and moves it into the destination dictionary.
+
+    If an object in the destination follows the stateful protocol, it loads the source state via ``load_state_dict``.
+
+    """
+    keys = set(source.keys()) if keys is None else keys
```
Comment (carmocca) [inline src/lightning/fabric/utilities/load.py:216]: This is the same
```suggestion
    keys = set(source) if keys is None else keys
```

Or accessing the keys should work too

```suggestion
    keys = source.keys() if keys is None else keys
```

File: src/lightning/fabric/utilities/load.py [hunk 3]
```diff
@@ -203,3 +203,21 @@ def _load_tensor(t: _NotYetLoadedTensor) -> Tensor:
...
+    for key in keys:
+        if key not in source:
+            continue
```
Comment (carmocca) [inline src/lightning/fabric/utilities/load.py:219]: nit: you could get the intersection of keys only in the keys != None case, so you avoid having to check this in each iteration

File: src/lightning/fabric/utilities/load.py [hunk 4]
```diff
@@ -203,3 +203,21 @@ def _load_tensor(t: _NotYetLoadedTensor) -> Tensor:
...
+        if key in destination and isinstance(destination[key], _Stateful):
+            destination[key].load_state_dict(source.pop(key))
+        else:
+            destination[key] = source.pop(key)
```
Comment (carmocca) [inline src/lightning/fabric/utilities/load.py:223]: nit: source.pop is called in both branches, you could move it out

File: tests/tests_fabric/strategies/test_deepspeed_integration.py [hunk 5]
```diff
@@ -392,6 +392,7 @@ def test_deepspeed_init_module_with_stage_3(empty_init):
     with mock.patch("deepspeed.zero.Init") as zero_init_mock, fabric.init_module(empty_init=empty_init):
         BoringModel()
     zero_init_mock.assert_called_once_with(enabled=True, remote_device=None, config_dict_or_path=ANY)
+    fabric.barrier()
```
Comment (carmocca) [inline tests/tests_fabric/strategies/test_deepspeed_integration.py:395]: why is this necessary? This is a standalone test so if there's a race condition here, perhaps the barrier should go `init_module` for DeepSpeed?

File: src/lightning/fabric/utilities/load.py [hunk 6]
```diff
@@ -203,3 +203,21 @@ def _load_tensor(t: _NotYetLoadedTensor) -> Tensor:
         return t._load_tensor()
 
     return apply_to_collection(collection, dtype=_NotYetLoadedTensor, function=_load_tensor)
+
+
+def _take_state_and_load_stateful(
```
Comment (carmocca) [inline src/lightning/fabric/utilities/load.py:208]: nit: i find this name odd: what about `_load_state_into`  or `_combine_state_with` ...?

**Review by carmocca — APPROVED** [review]

File: src/lightning/fabric/utilities/load.py (hunk 6 above)
Comment (awaelchli) [reply src/lightning/fabric/utilities/load.py:208]: wanted to express that the elements get removed from the source dict, so "take" was more expressive of that vs. "load".

File: tests/tests_fabric/strategies/test_deepspeed_integration.py (hunk 5 above)
Comment (awaelchli) [reply tests/tests_fabric/strategies/test_deepspeed_integration.py:395]: This is unrelated to the current PR changes sorry, but the test is flaky so I just tried this to see if it helps.

File: src/lightning/fabric/utilities/load.py (hunk 6 above)
Comment (carmocca) [reply src/lightning/fabric/utilities/load.py:208]: What about `move_state_into`?

File: tests/tests_fabric/strategies/test_deepspeed_integration.py (hunk 5 above)
Comment (carmocca) [reply tests/tests_fabric/strategies/test_deepspeed_integration.py:395]: I posted this comment because I don't see how the barrier would help. If you are referring to this error: https://dev.azure.com/Lightning-AI/lightning/_build/results?buildId=173794&view=logs&j=3f274fac-2e11-54ca-487e-194c91f3ae9f&t=8e4ceb7c-ceed-5ee0-b6fc-c9023e41cb74&l=597, it seems to happen on `launch()`

Comment (awaelchli) [reply tests/tests_fabric/strategies/test_deepspeed_integration.py:395]: It's easy to point to the line where the error happens and just say "here is the problem". But in this case it is very likely that the cause of the issue is somewhere else. The error you see is just what surfaces. Since this is not reproducible standalone locally, it is likely an interaction between test. With this guess of adding a barrier, I tried to eliminate the possibility that the two parameterizations run in sequence and don't leave enough time for the process group to be destroyed before the next init call. 

I was just trying to improve the developer experience with the test suite here. Fine with me, I'll remove it from here someone else will probably fix it in the future :)

Comment (awaelchli) [inline tests/tests_fabric/strategies/test_deepspeed_integration.py:395]:
```suggestion
```

**Review by Borda — APPROVED** [review]
